### ISSUE TITLE
Hostname verification for HTTPS URIs in twisted.web.client.Agent.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4888]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#4888|
|Type|defect|
|Created|2011-02-16 22:50:37Z|
|Branch|https://github.com/twisted/twisted/tree/redux-4888|

When we [ticket:4023 implemented HTTPS for the new client API], we left out any form of hostname verification or certificate checking.

We should support it in some way.

<details><summary>Searchable metadata</summary>

```
trac-id__4888 4888
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__Twisted_14_0_0 Twisted-14.0.0
branch__branches_redux_4888 branches/redux-4888
branch_author__hynek__itamarst__dreid__glyph hynek, itamarst, dreid, glyph
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1297896637000000 1297896637000000
changetime__1398385818000000 1398385818000000
version__None None
owner__glyph glyph
cc__ivank cc__hynek cc__Alex cc__Julian
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

First we need to support just checking that the hostname matches as expected.  Does Agent do that already?  Then, we need to actually verify the peer's certificate.

My inclination would be, by default, to support it opportunistically, verifying when possible, emitting a warning when it's not possible.

There should be separate tickets for supporting each of: the MacOS X certificate store, the Linux conventional default (/etc/ssl/certs) , and the Windows thing, which there are all references to on [#4023](https://github.com/twisted/twisted/issues/4023).  This ticket could probably be for the Linux default.  Or, maybe just for hostname checking against the (unverified) cert, if we don't do that yet.

There should be some kind of option, which says "for real, be secure, don't transmit my request in the clear, don't accept a response in the clear".  In my opinion, this flag should be specified independently of the protocol, since maybe there could be other transports besides HTTPS that could be secure.  Unfortunately for compatibility, this is probably what users mean when they take the care to type "https://" into a box that takes an URL.

I expect that there will be some further discussion first though.  Thoughts?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

> Or, maybe just for hostname checking against the (unverified) cert, if we don't do that yet.

I think this is a good idea.  It's not terribly useful to verify hostnames without verifying certificates, but the implementation of those two things are fairly independent.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

There's a [relevant RFC, 6125](http://tools.ietf.org/html/rfc6125) that an implementor of this feature should be aware of.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

See also [#5446](https://github.com/twisted/twisted/issues/5446) and [#5445](https://github.com/twisted/twisted/issues/5445) for related tasks.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

(In [37200]) Branching to 'hostname-verification-4888'

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_7"></a>|@radix commented|
|-|-|

This is a good feature, but we shouldn't have our own certificate database.

 * It's a liability to us in that we'd have to maintain it, even if we generate it from mozilla's DB. curl maintainers learned this the hard way. so are python-requests maintainers.

and it's a liability to our users:

 * the certificate store would become out of date (probably without them even knowing it)
 * administrators don't have management tools that work with this database included in Twisted

You should load certificates out of /etc/ssl/certs if it exists. If you want to solve this for users, then maybe include certs in the windows .exe distribution, but this shouldn't go into Twisted upstream.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_8"></a>|@radix commented|
|-|-|

On top of that, why is the CA bundle being added in this branch? it seems like it should be a part of a different ticket and not be bound to HTTPS hostname verification.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Replying to [radix](#note_8):
> On top of that, why is the CA bundle being added in this branch? it seems like it should be a part of a different ticket and not be bound to HTTPS hostname verification.

+1 to that. This should be split into as many discrete actions as possible, and shipping a cert bundle is obviously the most contentious.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

Replying to [radix](#note_7):
> This is a good feature, but we shouldn't have our own certificate database.
> 
>  * It's a liability to us in that we'd have to maintain it, even if we generate it from mozilla's DB. curl maintainers learned this the hard way. so are python-requests maintainers.
> 
> and it's a liability to our users:
> 
>  * the certificate store would become out of date (probably without them even knowing it)
>  * administrators don't have management tools that work with this database included in Twisted
> 
> You should load certificates out of /etc/ssl/certs if it exists. If you want to solve this for users, then maybe include certs in the windows .exe distribution, but this shouldn't go into Twisted upstream.

I mostly agree, except for a few minor points:

Let's not the perfect be the enemy of the good here.  We should not be in the cert-bundle shipping business if we can avoid it, but an out-of-date CA bundle is still better security in many cases than having to fall back to cleartext.  For example, pip's reluctance to ship a CA bundle lead to years of everyone just downloading and executing stuff plaintext because security was too hard.  It would have been a lot better for everyone if they - and we - had just shipped a CA bundle that wasn't getting updates and then worked to figure out a way to update it later.  Thanks for the prompting though, because I noticed that [ticket:5446 the ticket for looking at the platform certificate store] didn't actually have any API references on it and therefore it might not have been as clear as it could have what I was proposing there.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

As to the implementation in the branch itself; looking at [the function implemented for matching certificate names](http://twistedmatrix.com/trac/changeset/37253#file0), I'm wondering why we have one function for extracting the names and one function for matching the names, rather than a matcher object that wraps the certificate object.

For example, RFC 6125 seems very cagey on the concept of wildcard certificates, and wish that they weren't a thing, conceding that support is necessary only for "backward compatibility with deployed infrastructure".  It seems quite likely to me that some future X509 certificate or certificate authority metadata would add a field saying "no wildcards please" or otherwise modify the matching algorithm, based on some information that can't be represented in a list of certificate names (plus wildcards) represented by the certificate.  I would like this API to be as abstract – and therefore as future-proof – as possible, and so have an API which is just like "here's an object, it matches a hostname or it doesn't, don't worry about how".

(Also always include epytext `@type` fields on all parameters, especially parameters representing strings of characters or bytes.)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_12"></a>|@itamarst commented|
|-|-|

I should note that I'm doing this for one of my clients; this branch is just my way of dumping my first pass somewhere public. My plan therefore is to first get this to the point where it works, even if it's not mergeable. Later, in my free time or with the help of others, I will split this up into multiple branches as necessary, address review comments, write documentation etc..

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst commented|
|-|-|

As far as shipping CA bundle: it's not ideal, but it's better than nothing (as glyph noted in the pip example.) My current thoughts are that we should use the platform cert database if possible, falling back to a bundled cert database that we refresh on every release (if not more frequenly). Initially it'd probably be easiest to only support some set of Linux distributions, but we could add better defaults for other platforms over time.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

Replying to [itamar](#note_12):
> I should note that I'm doing this for one of my clients; this branch is just my way of dumping my first pass somewhere public. My plan therefore is to first get this to the point where it works, even if it's not mergeable. Later, in my free time or with the help of others, I will split this up into multiple branches as necessary, address review comments, write documentation etc..

I thought something like this might be the case, since you seem to be loading up the branch with lots of code :).  I'm offering comments now so there isn't one gigantic dump of review feedback at the end - and I thank radix for doing the same.  Also, this is security code, and benefits from extra eyeballs even more than usual.

On that note, regarding [another changeset](http://twistedmatrix.com/trac/changeset/37264):

  1. `getCACertificates` is a crappy name, in no small part because it starts with "get".  Call it `authorityCertificates` or `certificateAuthorityCertificates` or `trustedRoots` or something.
  2. This shouldn't return `X509` objects directly.  They probably segfault less than we're used to, but we already have a perfectly reasonable wrapper, `twisted.internet.ssl.Certificate`.  Return one of those; if you need to turn it back into an `X509`, you can.
  3. Oh my goodness, do not tell distributors to _patch_ a _function_!  Make a plugin, aggregate the certificates from the plugins.
  4. Loading multiple certificates from a single file is independently useful of loading platform CAs; in particular I believe this could be helpful when loading trust chains.  That code should be in a different function.  `Certificate.manyFromFile` perhaps.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

> For example, pip's reluctance to ship a CA bundle lead to years of everyone just downloading and executing stuff plaintext because security was too hard. It would have been a lot better for everyone if they - and we - had just shipped a CA bundle that wasn't getting updates and then worked to figure out a way to update it later.

I _don't_ think we should actually be in the business of managing and distributing such a bundle, but it sounds like enough other people are that it'll probably happen.

So.  Since apparently pip, python-requests, and Twisted should all have been bundling their own private CA databases for years now, how about creating a new, shared, independent package that consists only of such a bundle and can be depended on by all of these projects?  Hopefully the numerous advantages of such an approach over Twisted (and each of those other projects, perhaps plus others) are obvious, but if not:

  * It can be updated more frequently than Twisted itself is updated.  This is fairly important.  If you want to stop trusting a compromised authority, you want to stop trusted them today, not three months from now.
  * It provides consistent trust behavior across multiple projects.  It is better if all HTTPS software on a host comes to the same conclusion about a particular certificate (ie, the system provides a CA database).  Failing that, the more software agrees the better.  Having Twisted and pip always agree is certainly preferable to allowing them to use desynchronized CA databases and come to different conclusions about potentially compromised hosts.
  * It is less work.  It is one database to maintain and distribute instead of N.
  * There are more people to help with the work.  Many more people use and care about pip than use and care about Twisted.
  * Platform vendors could potentially someday be some of the people helping.  For example, Ubuntu could start maintaining an Ubuntu-specific version of this package which just points at the exact same certificates as are included in the ca-certificates package.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_15):

> I _don't_ think we should actually be in the business of managing and distributing such a bundle, but it sounds like enough other people are that it'll probably happen.

I would prefer to avoid this as well, and you make some good points.  And we haven't even started talking about licensing issues!  But, in any case, as we keep saying, this ticket is _not_ about including a CA bundle or in fact any kind of trust configuration, but about hostname matching, so we should stop filling up the comments here.  I have opened [ticket:6334 a new ticket specifically about this problem], described in such a way to make my feelings clear (i.e. we should not do it, I'm not entirely convinced that [ticket:5446 the ticket for fixing this the right way] is actually sufficiently harder than resolving all these issues).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

`PLATFORM = object()` is pretty crummy.  At the very least, there's no excuse for `object()` any more because we now have twisted.python.constants, and you can have a symbolic placeholder that actually has a nice stringification.

But, it would also be handy to say "the platform, plus some additional trusted roots that I added myself".  Given that you're enumerating the full list of platform CA certs anyway, why not provide an object that offers at least an `__iter__` if not a plain list of certs?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

"standard" security rules and "well known" certificate authorities are both kind of weasel words, too.  Since there's no actual standard governing which CAs to use, maybe "`BrowserLikeContextFactory`" would be a better name for it? (`LockIconContextFactory`, ha ha?)

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_19"></a>|@dreid commented|
|-|-|

Ok, this work actually looks pretty great.  But I couldn't help but notice that `StandardWebContextFactory` does not:

 1. Document the level of verification it does (it only implies that it is similar to browsers) this makes making an informed decision about if the level of verification it does perform is suitable for your application just a little bit harder.
 2. Provide any mechanism for specifying the caCerts.  Seems like `StandardWebContextFactory.__init__` sould take a `caCerts` keyword argument and pass it to `OpenSSLCertificateOptions` with it defaulting to `PLATFORM`.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_20"></a>|@dreid commented|
|-|-|

(In [38700]) BrowserLikeContextFactory instead. Refs [#4888](https://github.com/twisted/twisted/issues/4888)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_21"></a>|@itamarst commented|
|-|-|

Note to anyone doing more work on this: some of the code needs to be moved to other tickets' branches. In particular the CA bundle packaging should be moved to [#6334](https://github.com/twisted/twisted/issues/6334). [#5446](https://github.com/twisted/twisted/issues/5446) covers using openssl's knowledge of CA database on POSIX systems, and there's a couple of tickets for using OS X and Windows native CAs.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_23"></a>|@hynek commented|
|-|-|

(In [41550]) Branching to hostname-verification-4888-2.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_24"></a>|@hynek commented|
|-|-|

(In [41557]) Merge forward hostname verification for CertificateOptions

refs [#4888](https://github.com/twisted/twisted/issues/4888)

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_25"></a>|@hynek commented|
|-|-|

FYI, the implementation/merge of the Agent features is completely blocked by [#5446](https://github.com/twisted/twisted/issues/5446).

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_26"></a>|@hynek commented|
|-|-|

Another FYI, we’ll need our buildbots to install with pyasn1 and pyasn1_modules for the test suite…that is a per buildbot setting, right? Ought I send an email to the mailing list?

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_27"></a>|@hynek commented|
|-|-|

So it occurred to me how to make this branch a) better and b) unblock this ticket.

 1. Long story short, I want BrowserLikeContextFactory to depend on certifi. It _is_ called BrowserLike, what would it make more browser-like than using the CAs from an actual browser?
 1. Also slightly unrelatedly, I dislike the current API of passing the hostname into Certificate options, I would change it into adding an instance method called `.withHostnameValidationFor('hostname.com')` which would copy the current CF and add the validation.

Anyone feeling the urge to stop me from either?  While 1. was born out of desperation, I like it a lot. Using system CAs for web validation is asking for pain (just today  I was unable to wget something off my page which has a perfectly valid SSL config including chain from an older FreeBSD box).

If there’s consensus that my approach to 1. is wrong, I’m hereby arguing against calling the new Agent-CF “BrowserLike”, because it isn’t. I also don’t consider waiting for [#6934](https://github.com/twisted/twisted/issues/6934) to implement this an option; I’ll personally port this code to the new APIs it might introduce, but don’t want to wait anymore.

Looking forward to your feedback.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_28"></a>|@hynek commented|
|-|-|

Let me quickly show you the API I have in mind:

```python
class BrowserLikeContextFactory(object):
    def __init__(self):
        from twisted.internet._sslverify import OpenSSLCertificateOptions
        self._cf = OpenSSLCertificateOptions(verify=True,
                                             caCerts=loadCertifiCAs())


    def getContext(self, hostname, port):
        return self._cf.withHostnameVerificationFor(hostname, port).getContext()
```

I _do_ know that the amount of Factories is identical but I like the API a lot more and it gives some wiggle room for optimizations.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_29"></a>|@hynek commented|
|-|-|

Sorry i forgot, please compare to https://github.com/twisted/twisted/blob/27b4422c2460937c84457a08ee7d7e3324d3d28e/twisted/web/client.py#L718 – the old branch is a bit laborious to find.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_30"></a>|@dreid commented|
|-|-|

What about just adding this to WebContextFactory instead of a new object?  (We do still want this to be the default right?)

What specifically do you dislike about constructing CertificateOptions in getContext?

If we had an API like you propose would we need the WebContextFactory where getContext takes a hostname at all?

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_31"></a>|@hynek commented|
|-|-|

  - Can that ever happen?
  - I’m totally giving up on this because I don’t want to delay this ticket any further (I find that API gross non-withstanding).
  - I’m sure I understand your last question.  I think different behaviors will always be necessary. Once [#5446](https://github.com/twisted/twisted/issues/5446) is close, we could add a SystemTrustWebContextFactory. The current WebContextFactory should be deprecated and renamed to “InsecureWebContextFactoryDoNoUse” instead.  Does my response fit your question? :)

Do you have any feels about my certifi idea?  If it gets smashed down, I think I’ll publish something like that on PyPI.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_32"></a>|@radix commented|
|-|-|

I disagree with the idea of making BrowserLikeContextFactory *depend* on certifi. I think it should default to the platform's SSL database.

Certifi is worse than the platform's certificates and worse than a browser's certificates, because it has no sensible way to do certificate management by an operator of the system. It should just be a fallback. OS X, Windows, and Ubuntu all provide tools or standard methods for manipulating the certificates for your platform. Those tools should all affect the behavior of Twisted applications.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_33"></a>|@hynek commented|
|-|-|

Well the goal is to do the right thing for people who can’t/won’t do all that by default (cf. discussion about hostname validation in Python 3 and Ruby’s standard SSL settings).

System trust stores are terrible and just lead to disabling verification. I literally had to run wget with “--no-check-certificate” just today because FreeNAS/FreeBSD has an outdated/none? trust store and it wouldn’t connect to a perfectly valid SSL host.  As soon as [#5446](https://github.com/twisted/twisted/issues/5446) gives us the tools, we can add a factory for those who care about such things but I’d like to be secure by default.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_34"></a>|@glyph commented|
|-|-|

Replying to [hynek](#note_33):
> Well the goal is to do the right thing for people who can’t/won’t do all that by default (cf. discussion about hostname validation in Python 3 and Ruby’s standard SSL settings).

+1, we all agree this is a good idea.

> System trust stores are terrible and just lead to disabling verification.

This is a reason to fix the system trust store, not for every application to become an ad-hoc provisioner of CA trust.

> I literally had to run wget with “--no-check-certificate” just today because FreeNAS/FreeBSD has an outdated/none? trust store and it wouldn’t connect to a perfectly valid SSL host.

Maybe your "SSL host" was signed by TURKTRUST and the platform isn't trusting it on purpose!  Maybe your host is configured to only trust intranet sites.  The point is that we must leave this in a user's hands or we are effectively trying to be a CA ourselves.

> As soon as [[#5446](https://github.com/twisted/twisted/issues/5446)](https://github.com/twisted/twisted/issues/5446) gives us the tools, we can add a factory for those who care about such things but I’d like to be secure by default.

So, the subject of this ticket is "hostname verification".  I understand the need for a complete system of interlocking mechanisms to establish a real trust path, but that's not the best way to deal with implementation.  Can this ticket, or a dedicated sub-ticket, please limit itself to implementing _just_ the rfc6125 rules, checking the host name and SANs and whatnot (given the need for ASN.1 parsing, that seems more than hard enough for one ticket), and leave all the unpleasant stickiness of trust rules to #5446 and its potential successors?  This ticket really doesn't need _any_ trust stuff in it in order to fulfill its mission.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_35"></a>|@glyph commented|
|-|-|

As it happens, there's already a ticket, [#5190](https://github.com/twisted/twisted/issues/5190).  That could be just a function that verifies an issuer and returns `True` or `False`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_36"></a>|@glyph commented|
|-|-|

Before landing, this will also require [#6024](https://github.com/twisted/twisted/issues/6024), because that's the hook which will implement the appropriate place to _do_ the verification.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_37"></a>|@dreid commented|
|-|-|

(In [42121]) Branching to agent-contextFactory-hostname-4888.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_38"></a>|@glyph commented|
|-|-|

I'm reclassifying this as a defect.  In this case, I think the classification is actually pretty important, because if this were an "enhancement" then it should be implemented in a "compatible" way - i.e. just keep not verifying certificates for HTTPS servers unless you specifically ask for it.  As it is I think it falls under the "[gross violation of specifications](https://twistedmatrix.com/trac/wiki/CompatibilityPolicy#FixingGrossViolationofSpecifications)" clause of the compatibility policy, although it's in a bit of a grey area because it possibly fails the "published" part of the published specification test (hostname verification is RFC6125, but as far as I know there's no specific RFC specifying the trust cartel), so I will expound.  Perhaps there should be a separate explicit exception for security issues.

I think that it is conventionally well-understood that when a user types an `https` URI, either an application programmer typing it somewhere in their code, or a user entering it on the command line, they have the expectation that the communication will be verified with traditional CA-cartel trust.  More specialized applications might have _stricter_ trust, but it's misleading to have `https` in a URL that is totally unverified.

To be clear, this is an explicit repudiation of [comment:1 my earlier comment]:

> There should be some kind of option, which says "for real, be secure, don't transmit my request in the clear, don't accept a response in the clear".

There shouldn't be an option.  If anything, we should backport this to earlier point releases of Twisted, as it's a significant security fix; by allowing totally untrusted, unverified https connections to masquerade as "secure" from the users's perspective, I think that we were propagating a common security bug.

Thanks to all the security people who have taken the time to educate me on this issue, and who have help me understand the operational value of cartel trust despite the fact that it's not perfect.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_40"></a>|@glyph commented|
|-|-|

Please review [#7098](https://github.com/twisted/twisted/issues/7098) and see the comment there.  I can split up the changes if necessary.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_41"></a>|@glyph commented|
|-|-|

(In [42512]) Branching to redux-4888.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_42"></a>|@glyph commented|
|-|-|

Okay uh I guess this was already in review but now it's _actually_ in review.  Buildbot tells me there are some docstring fixes left to do, and a test skip thing, but other than that, would a reviewer like to opine?

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_43"></a>|Alex commented|
|-|-|

The docstring for `creatorForNetloc` includes a parameters which doesn't exist. Also, the `port` parameter is unused, that's ok? Since this is a new interface, is there any reason to include it? The interface `contextFactory` is not documented, is passing something besides `WebClientConnectionCreatorCreator` supported?

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_44"></a>|@wallrj commented|
|-|-|

A few observations from playing with the latest branch.

 1. What if you want to ignore the hostname mismatch in the certificate.  ie to
    download buildbot.twistedmatrix.com pages.
    * https://gist.github.com/wallrj/9174fb2b76e351a1e939

 2. And if people currently have scripts which rely on Agent's lack of hostname
    verification, won't it be a surprise if those suddenly start raising
    VerificationError.

 3. I also note that the VerificationError reported is in a private
    service_identity module and wrapped in another private web failure. Which
    makes it very difficult to handle and very difficult to work out which
    public exception to trap in an errback.

    ```
    (Twisted)[richard@zorin redux-4888]$ python docs/web/examples/https.py https://buildbot.twistedmatrix.com/json/builders/?as_text=1
    main function encountered error
    Traceback (most recent call last):
    Failure: twisted.web._newclient.ResponseNeverReceived: [<twisted.python.failure.Failure <class 'service_identity._common.VerificationError'>>]
    ```

    Is there any way to make that refer to the proper public APIs

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_45"></a>|@glyph commented|
|-|-|

Replying to [rwall](#note_44):
> A few observations from playing with the latest branch.
> 
>  1. What if you want to ignore the hostname mismatch in the certificate.  ie to
>     download buildbot.twistedmatrix.com pages.
>     * https://gist.github.com/wallrj/9174fb2b76e351a1e939

  1. Don't do that.
  2. Seriously, don't do that.  Tom should fix buildbot.  it's broken.  I made him a cert, like, a year ago, and maybe the fact that we will _care_ that it's broken might get us to fix it sooner.  (Actually he's probably going to fix it as soon as he gets back from vacation.)
  3. Make your own `CreatorCreator` which explicitly does the hostname translation.  Buildbot has a valid cert, it's just for the wrong domain; so pretend you're connecting to twistedmatrix.com at the SSL layer.
  4. Use plaintext HTTP.
  5. Did I mention you shouldn't do that?
  6. Make your own `CreatorCreator` which just returns a thing which creates a `CertificateOptions` and doesn't bother to verify anything.  And then feel bad because, as I mentioned, you shouldn't have done that.

>  2. And if people currently have scripts which rely on Agent's lack of hostname
>     verification, won't it be a surprise if those suddenly start raising
>     VerificationError.

And I'm sure there were lots of great bits of automation for reading the core memory of other servers on the internet that were all broken by the fact that we patched heartbleed.  Breaking this kind of stuff is sort of the _point_ of having security.

>  3. I also note that the VerificationError reported is in a private
>     service_identity module and wrapped in another private web failure. Which
>     makes it very difficult to handle and very difficult to work out which
>     public exception to trap in an errback.
...
>     Is there any way to make that refer to the proper public APIs

Yeah, this is a general badness of Agent, not really related to this change.  See [#5310](https://github.com/twisted/twisted/issues/5310) for fixing it.

Thanks for the feedback.  Looks like I'm nothing going to deal with in this batch, but it's good to make it explicit.  Looking forward to a real review though ;).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_46"></a>|@glyph commented|
|-|-|

Replying to [Alex](#note_43):
> The docstring for `creatorForNetloc` includes a parameters which doesn't exist. Also, the `port` parameter is unused, that's ok? Since this is a new interface, is there any reason to include it? The interface `contextFactory` is not documented, is passing something besides `WebClientConnectionCreatorCreator` supported?

Good points.  I'll fix those in the branch.

Anyone want to do a real review though? :)

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_47"></a>|@hynek commented|
|-|-|

FWIW, I agree with Rich we should ship a `IAcknowledgeThatIWillBeMITMedByUsingThisCreatorCreator` for internal systems etc.  We don’t want to add barriers for using Twisted in internal/legacy systems because that’s where it often shines.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_48"></a>|@exarkun commented|
|-|-|

> Don't do that.

"Sorry you can't interoperate with a random, large subset of the stupidly misconfigured part of the internet (by the way this is so hard to do *we* don't even do it right for *our* part of the internet)" sounds kind of lame to me.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_49"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_48):
> > Don't do that.
> 
> "Sorry you can't interoperate with a random, large subset of the stupidly misconfigured part of the internet

I never said that.  What "don't do that" means is that you _shouldn't_ interoperate with that misconfigured part of the internet; that's what "misconfigured" means, after all ;-).

Step 3 is the "right" answer here.  Once you know what hostname and certificate you're expecting, you can easily configure Twisted to verify _that_; it's just slightly harder (but still possible!) to tell it "don't verify this at all, I don't care about security at all".

> (by the way this is so hard to do *we* don't even do it right for *our* part of the internet)"

The difficulty of maintaining Twisted's infrastructure is not a great yardstick for the ease of maintainer infrastructure generally.  If this were the _only_ example my only answer would have been "fix your stuff, like we should fix ours"; since I know other people do also make this mistake, there are other options.

> sounds kind of lame to me.

There are plenty of options, but they should be slightly hard to discover, because we really really do not want anyone doing this by accident.  The first recourse here should always be to fix the infrastructure, and only once that's definitively not an option should you insert workarounds and hacks.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_50"></a>|@glyph commented|
|-|-|

Replying to [hynek](#note_47):
> FWIW, I agree with Rich we should ship a `IAcknowledgeThatIWillBeMITMedByUsingThisCreatorCreator` for internal systems etc.  We don’t want to add barriers for using Twisted in internal/legacy systems because that’s where it often shines.

If you want to add this, I strongly think it should be a separate ticket; the infrastructure is all present here to implement such a thing.

If it were possible to imply to someone using documentation or naming that a thing was ugly and shouldn't be used (hi there, `InMemoryUsernamePasswordDatabaseDontUse`) then what I'd say is that we should make it an example and not expose it as a public name anywhere, and actually require people to copy and paste the no-verification thing into their own code.  (But what that'll probably do is just make people copy a buggy example we can't ship security updates for, so...)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_51"></a>|@glyph commented|
|-|-|

Also for those of you concerned about `...CreatorCreator`, I think I'm going to rename that thing to `HTTPSPolicy`.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_52"></a>|@wallrj set owner to @glyph|
|-|-|

Glyph, here's a quick and dirty review while I wait for a flight. I'll try and make some more thorough notes later (unless someone gets there first)

Notes:
 * Bogus pydoctor OpenSSL reference error: https://buildbot.twistedmatrix.com/builders/documentation/builds/4677/steps/api-documentation/logs/new%20pydoctor%20errors
 * Merges cleanly
 * Coverage appears complete

Points:
 1. [branches/redux-4888/twisted/web/client.py](branches/redux-4888/twisted/web/client.py)
    1. Missing epydocs: https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/2047/steps/run-twistedchecker/logs/new%20twistedchecker%20errors
    2. The CreatorCreator suffix seems very confusing (as noted earlier) your suggested name seems better.

 2. [branches/redux-4888/twisted/web/test/test_agent.py](branches/redux-4888/twisted/web/test/test_agent.py)
    1. Build Failures due to OpenSSL import errors - please fix
       1. https://buildbot.twistedmatrix.com/builders/debian-easy-no-zope-py2.6-epoll/builds/1862/steps/epoll/logs/problems
       2. https://buildbot.twistedmatrix.com/builders/debian-easy-old-zope-py2.6-epoll/builds/1857/steps/epoll/logs/problems
       3. https://buildbot.twistedmatrix.com/builders/py-without-modules/builds/2456/steps/trial/logs/problems
       4. https://buildbot.twistedmatrix.com/builders/windows7-64-py2.7/builds/3872/steps/iocp/logs/problems
    2. Missing epydocs - please fix the important ones: https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/2047/steps/run-twistedchecker/logs/new%20twistedchecker%20errors
    3. Add a test for the deprecation of t.w.c.WebClientContextFactory
 3. Missing news file

Please answer or address the numbered points above and resubmit for review.

```
[ERROR]
Traceback (most recent call last):
  File "/home/buildslave/slaves/twisted/bot-idnar-debian64/easy-old-zope-2.6debian/venv/lib/python2.6/site-packages/Twisted-13.2.0-py2.6-linux-x86_64.egg/twisted/trial/runner.py", line 497, in loadPackage
    module = modinfo.load()
  File "/home/buildslave/slaves/twisted/bot-idnar-debian64/easy-old-zope-2.6debian/venv/lib/python2.6/site-packages/Twisted-13.2.0-py2.6-linux-x86_64.egg/twisted/python/modules.py", line 383, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "/home/buildslave/slaves/twisted/bot-idnar-debian64/easy-old-zope-2.6debian/venv/lib/python2.6/site-packages/Twisted-13.2.0-py2.6-linux-x86_64.egg/twisted/python/reflect.py", line 303, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "/home/buildslave/slaves/twisted/bot-idnar-debian64/easy-old-zope-2.6debian/venv/lib/python2.6/site-packages/Twisted-13.2.0-py2.6-linux-x86_64.egg/twisted/python/reflect.py", line 250, in _importAndCheckStack
    reraise(excValue, excTraceback)
  File "/home/buildslave/slaves/twisted/bot-idnar-debian64/easy-old-zope-2.6debian/venv/lib/python2.6/site-packages/Twisted-13.2.0-py2.6-linux-x86_64.egg/twisted/web/test/test_agent.py", line 40, in <module>
    from twisted.internet._sslverify import ClientTLSOptions
  File "/home/buildslave/slaves/twisted/bot-idnar-debian64/easy-old-zope-2.6debian/venv/lib/python2.6/site-packages/Twisted-13.2.0-py2.6-linux-x86_64.egg/twisted/internet/_sslverify.py", line 13, in <module>
    from OpenSSL import SSL, crypto, version
exceptions.ImportError: No module named OpenSSL
}}```{
[ERROR]
Traceback (most recent call last):
  File "/home/buildslave/slaves/twisted/bot-idnar-debian64/easy-old-zope-2.6debian/venv/lib/python2.6/site-packages/Twisted-13.2.0-py2.6-linux-x86_64.egg/twisted/trial/runner.py", line 497, in loadPackage
    module = modinfo.load()
  File "/home/buildslave/slaves/twisted/bot-idnar-debian64/easy-old-zope-2.6debian/venv/lib/python2.6/site-packages/Twisted-13.2.0-py2.6-linux-x86_64.egg/twisted/python/modules.py", line 383, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "/home/buildslave/slaves/twisted/bot-idnar-debian64/easy-old-zope-2.6debian/venv/lib/python2.6/site-packages/Twisted-13.2.0-py2.6-linux-x86_64.egg/twisted/python/reflect.py", line 303, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "/home/buildslave/slaves/twisted/bot-idnar-debian64/easy-old-zope-2.6debian/venv/lib/python2.6/site-packages/Twisted-13.2.0-py2.6-linux-x86_64.egg/twisted/python/reflect.py", line 250, in _importAndCheckStack
    reraise(excValue, excTraceback)
  File "/home/buildslave/slaves/twisted/bot-idnar-debian64/easy-old-zope-2.6debian/venv/lib/python2.6/site-packages/Twisted-13.2.0-py2.6-linux-x86_64.egg/twisted/web/test/test_agent.py", line 40, in <module>
    from twisted.internet._sslverify import ClientTLSOptions
  File "/home/buildslave/slaves/twisted/bot-idnar-debian64/easy-old-zope-2.6debian/venv/lib/python2.6/site-packages/Twisted-13.2.0-py2.6-linux-x86_64.egg/twisted/internet/_sslverify.py", line 13, in <module>
    from OpenSSL import SSL, crypto, version
exceptions.ImportError: No module named OpenSSL
}}

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_53"></a>|@glyph commented|
|-|-|

Thanks, rwall!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_55"></a>|@glyph commented|
|-|-|

(In [42535]) Fix deprecation warning strings.

refs [#4888](https://github.com/twisted/twisted/issues/4888)
refs [#7242](https://github.com/twisted/twisted/issues/7242)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_56"></a>|@glyph removed owner|
|-|-|

Replying to [rwall](#note_52):
> Glyph, here's a quick and dirty review while I wait for a flight. I'll try and make some more thorough notes later (unless someone gets there first)
> 
> Points:
>  1. [branches/redux-4888/twisted/web/client.py](branches/redux-4888/twisted/web/client.py)
>     1. Missing epydocs: https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/2047/steps/run-twistedchecker/logs/new%20twistedchecker%20errors

Aside from these, there are spurious twistedchecker errors yet again.

I filed https://github.com/twisted/twistedchecker/issues/76 for `set_connect_state` (which is the only unfixable one) and https://github.com/twisted/twistedchecker/issues/75 for the inner-class stuff which I've worked around.

>     2. The CreatorCreator suffix seems very confusing (as noted earlier) your suggested name seems better.

Right-o.  Name changed.

>  2. [branches/redux-4888/twisted/web/test/test_agent.py](branches/redux-4888/twisted/web/test/test_agent.py)
>     1. Build Failures due to OpenSSL import errors - please fix
>        1. https://buildbot.twistedmatrix.com/builders/debian-easy-no-zope-py2.6-epoll/builds/1862/steps/epoll/logs/problems
>        2. https://buildbot.twistedmatrix.com/builders/debian-easy-old-zope-py2.6-epoll/builds/1857/steps/epoll/logs/problems
>        3. https://buildbot.twistedmatrix.com/builders/py-without-modules/builds/2456/steps/trial/logs/problems
>        4. https://buildbot.twistedmatrix.com/builders/windows7-64-py2.7/builds/3872/steps/iocp/logs/problems

These all seem to be the same failure.  Perhaps more of our build farm should have pyOpenSSL installed?  There's really just supposed to be the one builder without optional dependencies, I thought.

>     2. Missing epydocs - please fix the important ones: https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/2047/steps/run-twistedchecker/logs/new%20twistedchecker%20errors

I think these should all be fixed now, important or not :).

>     3. Add a test for the deprecation of t.w.c.WebClientContextFactory

Hmm.  While I'm at it, I think this should probably be using deprecatedModuleAttribute, actually; at least until the tickets referenced by ticket:4804#comment:15 are addressed.

And then when I tried to do the right thing here I discovered [#7242](https://github.com/twisted/twisted/issues/7242) as well, and filed it; you can see the comment in the test code.

>  3. Missing news file

Added.

I also added a formal interface because all these silly half-finished
looks-kinda-like-this-one-weird-class types are a bad idea and I should stop
putting them into public interfaces.

Last time when I put it up for review I knew it was in kinda bad shape and I
wanted feedback before putting more effort in; this time I think it's probably
ready to go.

I also filed [#7243](https://github.com/twisted/twisted/issues/7243) for other APIs.  Hopefully that's a duplicate of something?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_57"></a>|@exarkun commented|
|-|-|

It seems this ticket has been de facto promoted to a regression (it is blocking the 14.0 release because ... "reasons" ... I don't know).  Adjusting metadata to reflect this unfortunate reality.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_58"></a>|@glyph commented|
|-|-|

While I'd like this to get into the 14.0 release (we're so close, and this is such an important fix!) this does _not_ actually need to block it.  If necessary we can have a quick follow-up 14.1 that closes this issue - which was the original plan anyway.

[#7098](https://github.com/twisted/twisted/issues/7098) did, because it involved a pretty significant re-working of an unreleased public interface, rolling back and re-implementing would have been a ton of additional work, and it changed around a bunch of public exposure of guts which had already landed on trunk which I did not want us to have to live with long term.  But that's closed now and nothing in the branch for this ticket needs to change any of those interfaces, so we're OK to have a release that does not include this.

Randomly changing around metadata is fun so I've done that as well.  I've put it in the milestone in case we do want to block but I am explicitly leaving this at hawkowl's discretion to wait on or not.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_59"></a>|@dreid commented|
|-|-|

From 4888.bugfix

```
twisted.web.client.Agent now verifies 'https' URLs against (its best guess of)
the platform trust roots.
```

According to the previous docstring for ```WebClientContextFactory``` it was already doing basic certificate checking by passing ```platformRoot``` as ```trustRoot```.  The news fragment should probably mention what this branch actually does which is hostname
verification.

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_60"></a>|Alex set owner to @glyph|
|-|-|

Review:

* `_ContextFactoryWithContext` this class should not be created as a closure, it should be a normal top level class which takes a parameter to its constructor. This will be considerably faster on both PyPy and CPython, as well as requiring less memory.

* ` A L{HTTPConnectionPool}` this should be "an"

* `contextFactory = _DeprecatedToCurrentPolicyForHTTPS(contextFactory)` a deprecation warning should be emitted when this is hit.

Please fix these items, and then it looks good to me.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_61"></a>|@glyph commented|
|-|-|

Replying to [Alex](#note_60):
> Review:
> 
> * `_ContextFactoryWithContext` this class should not be created as a closure, it should be a normal top level class which takes a parameter to its constructor. This will be considerably faster on both PyPy and CPython, as well as requiring less memory.

Sigh, where are JavaScript object literals when you need them…

Sure.

> * ` A L{HTTPConnectionPool}` this should be "an"

Good catch.  I should probably make it easier for myself to read the generated API docs as I'm writing them.

> * `contextFactory = _DeprecatedToCurrentPolicyForHTTPS(contextFactory)` a deprecation warning should be emitted when this is hit.

Oh, that's a good point.  I deprecated the other thing, but I guess if you provide something that's _like_ a `WebClientContextFactory` but doesn't actually _import_ a `WebClientContextFactory` you won't see the deprecation.

> Please fix these items, and then it looks good to me.

Thanks for the review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_62"></a>|@glyph commented|
|-|-|

OK, feedback dealt with, I think.  Double-checking with buildbot right now, then landing.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_63"></a>|@glyph set status to `closed`|
|-|-|

(In [42546]) Merge redux-4888

Author: glyph, dreid, hynek
Reviewer: alex, rwall
Fixes: [#4888](https://github.com/twisted/twisted/issues/4888)

Make https work right.
